### ISSUE TITLE
How order of signatures is ensured using ld-signatures?

### ISSUE THREAD

**Issue by Drabiv** [w3c-dvcg/ld-signatures#17]
When there are multiple signatures on a document, it is often important to have a proof of the order of signing. For example, signature of a notary, signifies that notary witnessed signing of a document by  signer.

How is order of signatures ensured, prooved when using ld-signatures? There is a section about signature chains - https://w3c-dvcg.github.io/ld-signatures/#signature-chains, but it is not clear to me how to verify that the order of signatures was not changed?
Does the signature value of 2nd signature made by signing hash made of Message + 1st signature? 

What is the algorithm to place signatures in order and to verify order of signatures?

**Comment by msporny** [comment]
> Does the signature value of 2nd signature made by signing hash made of Message + 1st signature?

Yes, the first signature is included in the normalized data that is hashed and signed by the 2nd signature.

> What is the algorithm to place signatures in order and to verify order of signatures?

It's basically a for loop that starts with the array of signatures. The algorithm then creates an object containing only the first signatures, then verifies it. If the signature is valid, it adds the second signature, then verifies it... and so on until all signatures in the chain are verified. Every subsequent signature verifies that the previous signature was included in the signature.

**Comment by Drabiv** [comment]
Thanks @msporny, that clarifies it.
I have a follow up question. When there are 3 and more signatures, then there are many ways in which signatures may have been put. How is this handled?
For example, if there are 3 signatures, the 3rd signature may have signed "Message+Sig#1+Sig#2" or it may have signed "Message+Sig#1" or "Message+Sig#2". Is there a way to convey what signatures has been signed by the signature?

**Comment by msporny** [comment]
> if there are 3 signatures, the 3rd signature may have signed "Message+Sig#1+Sig#2" or it may have signed "Message+Sig#1" or "Message+Sig#2". Is there a way to convey what signatures has been signed by the signature?

Note that we're changing the spec to use "proof" instead of "signature". I'll be using the "proof" language below.

If the signatures reside in "proof" array, then the assumption is that it's a set of signatures and NO signature was included in the hash of the other signatures. If the signatures reside in a "proofChain" array, then the assumption is that every previous signature was included in the current signature.

At present, there is no way to mix and match certain signatures counter-signing other signatures. We just don't have a popular use case for such a thing.

**Comment by Drabiv** [comment]
@msporny thanks, that's clear now.

**Comment by msporny** [comment]
Re-opening as I don't think the spec is clear on this and we don't specify the algorithms I outline above. The reference implementations do this, I think, but the spec should state it clearly as well.

**Comment by cwebber** [comment]
Indeed, I was just looking today for the text... the spec currently [introduces signature chains and signature sets](https://w3c-dvcg.github.io/ld-signatures/#multiple-signatures) but does not describe either of them.  It seems that the behavior both in signing and in verifying in both cases is a bit different depending on whether signature chain or signature set is selected, so we should probably write language for both.
